### ISSUE TITLE
k8s 1.6: workload endpoint deleted though Pod is still running

### ISSUE THREAD

**Issue by r0bj** [projectcalico/calicoctl#1592]
kubernetes: 1.6.1
calico: 2.1
docker: 1.12.6
OS: ubuntu 14.04

calico k8s config from http://docs.projectcalico.org/v2.1/getting-started/kubernetes/installation/hosted/calico.yaml:
https://gist.github.com/r0bj/26e77c1698bb9bfb0796924fba600a1f

After k8s node restart sometimes workload endpoints are not created properly. In the example below there is one pod which is broken in terms of network communication.

```
# docker ps
CONTAINER ID        IMAGE                                                                                                                            COMMAND                  CREATED             STATUS              PORTS               NAMES
b17f0797faef        gcr.io/google_containers/pause-amd64:3.0                                                                                         "/pause"                 5 minutes ago       Up 5 minutes                            k8s_POD_traefik-ingress-controller-h2w1q_kube-system_18ecd7d3-192b-11e7-bb29-5254008a1481_4
```

So there is container `k8s_POD_traefik-ingress-controller-h2w1q_kube-system_18ecd7d3-192b-11e7-bb29-5254008a1481_4`.

```
# docker inspect --format '{{.State.Pid}}' b17f0797faef
2455

# nsenter -t 2455 -n ip addr show
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: tunl0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN group default qlen 1
    link/ipip 0.0.0.0 brd 0.0.0.0
4: eth0@if6: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default
    link/ether 02:57:ca:8e:9a:9a brd ff:ff:ff:ff:ff:ff
    inet 10.203.10.142/32 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::57:caff:fe8e:9a9a/64 scope link
       valid_lft forever preferred_lft forever
```
Pod address is `10.203.10.142`.

Routing table on node:
```
# ip r
default via 10.14.30.1 dev eth0
10.14.30.0/24 dev eth0  proto kernel  scope link  src 10.14.30.124
10.203.5.192/26 via 10.14.30.111 dev tunl0  proto bird onlink
blackhole 10.203.10.128/26  proto bird
10.203.72.64/26 via 10.14.30.117 dev tunl0  proto bird onlink
10.203.98.0/26 via 10.14.30.116 dev tunl0  proto bird onlink
10.203.100.192/26 via 10.14.30.110 dev tunl0  proto bird onlink
10.203.137.64/26 via 10.14.30.112 dev tunl0  proto bird onlink
10.203.185.192/26 via 10.14.30.145 dev tunl0  proto bird onlink
10.203.219.0/26 via 10.14.30.118 dev tunl0  proto bird onlink
172.17.0.0/16 dev docker0  proto kernel  scope link  src 172.17.0.1
```

So there is no route for address `10.203.10.142` which is present inside failed pod.
There should be route like:
```
10.203.10.142 dev calib81f25735b1  scope link
```

```
# ip l
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP mode DEFAULT group default qlen 1000
    link/ether 52:54:00:b7:3c:92 brd ff:ff:ff:ff:ff:ff
3: docker0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN mode DEFAULT group default
    link/ether 02:42:9e:46:e5:ec brd ff:ff:ff:ff:ff:ff
4: tunl0@NONE: <NOARP,UP,LOWER_UP> mtu 1440 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1
    link/ipip 0.0.0.0 brd 0.0.0.0
5: cali1198573854e@tunl0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default
    link/ether be:6d:97:ab:59:4a brd ff:ff:ff:ff:ff:ff
6: calib81f25735b1@tunl0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default
    link/ether 92:7b:19:10:0d:22 brd ff:ff:ff:ff:ff:ff
```

```
# calicoctl get workloadEndpoint
NODE                ORCHESTRATOR   WORKLOAD                                            NAME
dev-k8s-worker-p2   k8s            kube-system.backend-healthcheck-3429883846-pmfmm    eth0
dev-k8s-worker-p2   k8s            kube-system.backend-healthcheck-3429883846-ts8db    eth0
dev-k8s-worker-p2   k8s            kube-system.default-http-backend-66l0w              eth0
dev-k8s-worker-p2   k8s            kube-system.heapster-3549184524-xzhtf               eth0
dev-k8s-worker-p2   k8s            kube-system.kube-dns-3487006785-p2d65               eth0
dev-k8s-worker-p2   k8s            kube-system.kube-dns-3487006785-rtk14               eth0
dev-k8s-worker-p2   k8s            kube-system.kube-dns-3487006785-x1j53               eth0
dev-k8s-worker-p2   k8s            kube-system.kube-dns-autoscaler-2815297830-6btl7    eth0
dev-k8s-worker-p2   k8s            kube-system.kubernetes-dashboard-1019458639-glbx7   eth0
dev-k8s-worker-p2   k8s            ops.grafana-2186705015-shk5m                        eth0
dev-k8s-worker-p2   k8s            ops.nginx-stateful-test-0                           eth0
dev-k8s-worker-p2   k8s            ops.nginx-stateful-test-1                           eth0
dev-k8s-worker-p2   k8s            ops.nginx-test-2647610131-lkkqn                     eth0
dev-k8s-worker-p2   k8s            ops.nginx-test-2647610131-ttx6m                     eth0
dev-k8s-worker-p3   k8s            default.backend-healthcheck-3631583507-t7ndj        eth0
dev-k8s-worker-p3   k8s            default.rychu-poc-2174780777-wnc88                  eth0
dev-k8s-worker-p3   k8s            kube-system.kube-dns-v20-1485703853-06wnl           eth0
dev-k8s-worker-p3   k8s            kube-system.kubernetes-dashboard-1655269645-j3r6c   eth0
```

There is no workload endpoint for failed container `kube-system.traefik-ingress-controller-h2w1q`.

So in this case kubernetes keeps restarting container inside pod but with no luck because the whole network namespace for this pod has broken communication.

`calico-node` logs:
https://gist.github.com/r0bj/be717f5885488542d04bd5e9d101f409

According to logs:
```
2017-04-04 12:53:22.595 [INFO][95] route_table.go 293: Syncing routes: adding new route. ifaceName="calib81f25735b1" ipVersion=0x4 targetCIDR=10.203.10.142/32
2017-04-04 12:53:22.596 [INFO][95] status_combiner.go 81: Endpoint up for at least one IP version id=proto.WorkloadEndpointID{OrchestratorId:"k8s", WorkloadId:"kube-system.traefik-ingress-controller-h2w1q", EndpointId:"eth0"} ipVersion=0x4 status="up"
2017-04-04 12:53:22.596 [INFO][95] status_combiner.go 98: Reporting combined status. id=proto.WorkloadEndpointID{OrchestratorId:"k8s", WorkloadId:"kube-system.traefik-ingress-controller-h2w1q", EndpointId:"eth0"} status="up"
2017-04-04 12:53:22.596 [INFO][95] int_dataplane.go 554: Finished applying updates to dataplane. msecToApply=5.740474000000001
```
So route is created.

But moment later for some reason endpoint is removed:
```
2017-04-04 12:53:44.634 [INFO][95] route_table.go 274: Syncing routes: removing old route. dest=10.203.10.142/32 ifaceName="calib81f25735b1" ipVersion=0x4
2017-04-04 12:53:44.634 [INFO][95] conntrack.go 67: Removing conntrack flows ip=10.203.10.142
2017-04-04 12:53:44.645 [INFO][95] status_combiner.go 86: Reporting endpoint removed. id=proto.WorkloadEndpointID{OrchestratorId:"k8s", WorkloadId:"kube-system.traefik-ingress-controller-h2w1q", EndpointId:"eth0"}
2017 
```

Restarting this pod fixes the issue.

Please feel free to ask If you need some additional info or data.

**Comment by caseydavenport** [comment]
@r0bj thanks for raising - looks like something is going wrong and the workload endpoint is being deleted.

I think it would be useful to have the full kubelet logs so we can see what is happening in CNI throughout the node reboot.

Could you enable debug logging in the Calico CNI plugin (`log_level: debug`) and verbose logging in the kubelet (`--v=5`) and then share those logs?

**Comment by r0bj** [comment]
@caseydavenport here you have:

calico-node logs:
https://gist.github.com/r0bj/ef15d5767b0ac3805f5c67bdbc610b8d

kubelet logs:
https://gist.github.com/r0bj/0a9ee90009a5721b38a6864ab62918bc

failed pod name: `traefik-ingress-controller-g6lsb`

**Comment by caseydavenport** [comment]
Investigating this now.  I have been able to reproduce on v1.6.

Not clear yet if this is a Calico bug, a Kubernetes bug, or somewhere in between, but marking as high priority.

**Comment by caseydavenport** [comment]
Latest investigation shows that the kubelet is calling the CNI plugin to teardown the Pod using the old container ID after the Pod has been restarted using a new container ID.

The Calico CNI plugin is respecting that call, and deleting the workloadendpoint for that Pod even though the Pod is still running.

This is because Calico uses the pod name/namespace as the identifying information for the workload endpoint rather than the container ID.  This was required to work around a bug in upstream Kubernetes, but is also desirable from an ease of use / debugging standpoint.

I've raised https://github.com/kubernetes/kubernetes/issues/44100 to attempt to address this upstream by ensuring Kubernetes doesn't call DEL after the Pod has been restarted.

**Comment by caseydavenport** [comment]
Fix being discussed in  projectcalico/cni-plugin#310

**Comment by immesys** [comment]
Is there a workaround? This is crippling me

**Comment by sigxcpu76** [comment]
This completely breaks StatefulSets because they use the same namespace/name upon restart.
The new pod will have no network few tens of seconds after startup.

What is the actual reasoning to identify a pod by its ```name.namespace``` instead of a real unique ID?

**Comment by immesys** [comment]
Yeah my statefulset pods don't get networking when I delete them and they are recreated. I have to scale down to zero instances, wait and scale back up to fix it

**Comment by tomdee** [comment]
Fixed with cni-plugin v1.7.0
